README.md
# minitest-spec-expect

[![Build Status](https://travis-ci.org/davejachimiak/minitest-spec-expect.png?branch=master)](https://travis-ci.org/davejachimiak/minitest-spec-expect)

Expect syntax for [Minitest](http://docs.seattlerb.org/minitest/index.html). Made out of love for
both rspec2's expect syntax and Minitest's lightweight feel.
## Install
In your Gemfile:
```ruby
gem 'minitest-spec-expect'
```
or from the command line:
```
$ gem install minitest-spec-expect
```
Then require it after your runner:
```ruby
require 'minitest/autorun'
require 'minitest/spec/expect'
```

minitest-spec-expect works in the following ruby versions and implementations:
* 1.9.2
* 1.9.3
* 2.0.0
* 2.1.0
* JRuby

## Usage
Wrap the object under test in an `expect()` object. Then call a Minitest expectation on it,
substituting `must` and `wont` with `to` and `to_not`.

For example:
```ruby
describe Integer do
  describe '#+' do
    it 'adds numbers correctly' do
      expect(1 + 1).to_equal 2
    end

    it 'does not add numbers incorrectly' do
      expect(1 + 1).to_not_equal 3
    end
  end
end
```
That's equivalent to:
```ruby
describe Integer do
  describe '#+' do
    it 'adds numbers correctly' do
      (1 + 1).must_equal 2
    end

    it 'does not add numbers incorrectly' do
      (1 + 1).wont_equal 3
    end
  end
end
```
## API
Please see the Minitest [expectation](http://docs.seattlerb.org/minitest/Minitest/Expectations.html)
docs for details on corresponding `must_*` and `wont_*` methods.
### `#*be`
```ruby
expect(1).to_be :<, 2
expect(1).to_not_be :>, 2
```
### `#*be_close_to/#*be_within_delta`
```ruby
expect(Math::PI).to_be_close_to 22.0/7.0, 0.01
expect(Math::PI).to_be_within_delta 22.0/7.0, 0.01
expect(Math::PI).to_not_be_close_to 22.0/6.0, 0.01
expect(Math::PI).to_not_be_within_delta 22.0/6.0, 0.01
```
### `#*be_empty`
```ruby
expect([]).to_be_empty
expect({awwyeah: :minitest}).to_not_be_empty
```
### `#*be_instance_of`
```ruby
expect([]).to_be_instance_of Array
expect([]).to_not_be_instance_of Hash
```
### `#*be_kind_of`
```ruby
expect([]).to_be_kind_of Enumerable
expect([]).to_not_be_kind_of NilClass
```
### `#*be_nil`
```ruby
expect(nil).to_be_nil
expect(:something).to_not_be_nil
```
### `#*be_same_as`
```ruby
expect(obj = Object.new).to_be_same_as obj
an_object = Object.new
expect(Object.new).to_not_be_same_as an_object
```
### `#to_be_silent`
```ruby
expect { '' }.to_be_silent
expect(->{''}).to_be_silent
```
### `#*be_within_epsilon`
```ruby
expect(4).to_be_within_epsilon 3.5, 0.15
expect(4).to_not_be_within_epsilon 3.5, 0.11
```
### `#*equal`
```ruby
expect('4d3d3d3d').to_equal '4d3d3d3d'
expect('4d3d3d3d').to_not_equal 'tayne'
```
### `#*include`
```ruby
expect(['tim', 'eric']).to_include 'eric'
expect(['tim', 'eric']).to_not_include 'brule'
```
### `#*match`
```ruby
expect('fartfartfartfart').to_match /(fart)*/
expect('bart').to_not_match /fart/
```
### `#to_output`
```ruby
expect { puts 'barf' }.to_output "barf\n"
expect(->{ puts 'barf' }).to_output "barf\n"
```
### `#to_raise`
```ruby
expect { raise RuntimeError }.to_raise RuntimeError
expect(->{ raise RuntimeError }).to_raise RuntimeError
```
### `#*respond_to`
```ruby
expect(Object.new).to_respond_to :must_respond_to
expect(Object.new).to_not_respond_to :fart_factory
```
### `#*throw`
```ruby
expect { throw StandardError }.to_throw StandardError
expect(->{ throw StandardError }).to_throw StandardError
```
## Contribute
1. Fork the repo.
2. Create a branch.
3. Code.
4. Ensure the specs are green (`$ rake`)
5. Open a pull request.

## License
The MIT License (MIT)

Copyright (c) 2013 Dave Jachimiak

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
#### version 2.0.0 - *July 10, 2014*
* Enforce that users require a runner. Via [@blowmage](https://github.com/blowmage).

#### version 1.0.0 - *December 28, 2013*
* Simplified initializer for expect object. Make stable.

#### version 0.1.0 - *May 23, 2013*
* Added ability to pass a block to `#expect`. When a block is passed to it, the expect object responds
to `#to_throw`, `#to_raise`, `#to_output`, and `#to_be_silent`. Big thanks to
[@splattael](https://github.com/splattael).


test.rb
old_instance_methods = Object.instance_methods

require 'minitest/spec'

$infected_assertions = Object.instance_methods - old_instance_methods

module Kernel
  def expect object
    Expect.new object
  end
end

class Expect
  attr_reader :object

  def initialize object
    @object = object
  end

  $infected_assertions.each do |assertion|
    method = assertion.to_s.gsub('must', 'to').gsub 'wont', 'to_not'

    self.class_eval <<-EOM
      def #{method} *args
        object.#{assertion} *args
      end
    EOM
  end
end

require 'minitest/autorun'

describe 'forty two' do
  before { @forty_two = 42 }

  it 'equals 42' do
    expect(@forty_two).to_equal 42
  end

  it 'is less than 43' do
    expect(@forty_two).to_be :<, 43
  end

  it 'is not more than 43' do
    expect(@forty_two).not_to_be :>, 43
  end

  it 'is a fixnum' do
    expect(@forty_two).to_be_instance_of Fixnum
  end

  it 'is a kind of integer' do
    expect(@forty_two).to_be_kind_of Integer
  end

  it 'is not a kind of float' do
    expect(@forty_two).to_not_be_kind_of Float
  end
end


spec/kernel_spec.rb
$:.unshift File.expand_path("../../lib", __FILE__)

require 'minitest/autorun'
require 'minitest/spec/expect'

describe Kernel do
  describe '#expect' do
    it 'explodes if nothing is passed to it' do
      ->{ expect }.must_raise ArgumentError, 'must pass an argument or a block'
    end

    it 'explodes if both an arg and a block passed to it' do
      ->{ expect(:straw_man) { :cinder } }.
        must_raise ArgumentError, 'cannot handle both an argument and a block'
    end

    describe 'when a block is passed' do
      before do
        expectations              = Minitest::Expectations.instance_methods
        block_expectation_regexes = [/silent/, /output/, /raise/, /throw/]

        non_block_expectations = expectations.reject do |method|
          block_expectation_regexes.detect do |regex|
            method.to_s.match regex
          end
        end

        transpositions = { 'must' => 'to', 'wont' => 'to_not' }

        @non_block_expect_methods = non_block_expectations.map do |expectation|
          transpositions.inject('') do |memo, transposition|
            string = memo.empty? ? expectation.to_s : memo

            string.gsub transposition.first, transposition.last
          end.to_sym
        end
      end

      it "won't respond to non-block-passing expectations" do
        @non_block_expect_methods.each do |method|
          expect { :muscular_son }.wont_respond_to method
        end
      end
    end
  end
end


spec/integration_spec.rb
$:.unshift File.expand_path("../../lib", __FILE__)

require 'minitest/spec/expect/autorun'

describe 'expect syntax' do
  it 'supports must_be as to_be' do
    expect(1).to_be :<, 2
  end

  it 'supports wont_be as to_not_be' do
    expect(1).to_not_be :>, 2
  end

  it 'supports must_be_close_to as to_be_close_to (within_delta)' do
    expect(Math::PI).to_be_close_to 22.0/7.0, 0.01
    expect(Math::PI).to_be_within_delta 22.0/7.0, 0.01
  end

  it 'supports wont_be_close_to as to_not_be_close_to (within_delta)' do
    expect(Math::PI).to_not_be_close_to 22.0/6.0, 0.01
    expect(Math::PI).to_not_be_within_delta 22.0/6.0, 0.01
  end

  it 'supports must_be_empty as to_be_empty' do
    expect([]).to_be_empty
  end

  it 'supports wont_be_empty as to_not_be_empty' do
    expect({awwyeah: :minitest}).to_not_be_empty
  end

  it 'supports must_be_instance_of as to_be_instance_of' do
    expect([]).to_be_instance_of Array
  end

  it 'supports wont_be_instance_of as to_not_be_instance_of' do
    expect([]).to_not_be_instance_of Hash
  end

  it 'supports must_be_kind_of as to_be_kind_of' do
    expect([]).to_be_kind_of Enumerable
  end

  it 'supports wont_be_kind_of as to_not_be_kind_of' do
    expect([]).to_not_be_kind_of NilClass
  end

  it 'supports must_be_nil as to_be_nil' do
    expect(nil).to_be_nil
  end

  it 'supports wont_be_nil as to_not_be_nil' do
    expect(:something).to_not_be_nil
  end

  it 'supports must_be_same_as as to_be_same_as' do
    expect(obj = Object.new).to_be_same_as obj
  end

  it 'supports wont_be_same_as as to_not_be_same_as' do
    object_one = Object.new

    expect(Object.new).to_not_be_same_as object_one
  end

  it 'supports must_be_silent as to_be_silent' do
    expect { '' }.to_be_silent
    expect(->{''}).to_be_silent
  end

  it 'supports must_be_within_epsilon as to_be_within_epsilon' do
    expect(4).to_be_within_epsilon 3.5, 0.15
  end

  it 'supports wont_be_within_epsilon as to_not_be_within_epsilon' do
    expect(4).to_not_be_within_epsilon 3.5, 0.11
  end

  it 'supports must_equal as to_equal' do
    expect('4d3d3d3d').to_equal '4d3d3d3d'
  end

  it 'supports wont_equal as to_not_equal' do
    expect('4d3d3d3d').to_not_equal 'tayne'
  end

  it 'supports must_include as to_include' do
    expect(['tim', 'eric']).to_include 'eric'
  end

  it 'supports wont_include as to_not_include' do
    expect(['tim', 'eric']).to_not_include 'brule'
  end

  it 'supports must_match as to_match' do
    expect('fartfartfartfart').to_match /(fart)*/
  end

  it 'supports wont_match as to_not_match' do
    expect('bart').to_not_match /fart/
  end

  it 'supports must_output as to_output' do
    expect { puts 'barf' }.to_output "barf\n"
    expect(->{ puts 'barf' }).to_output "barf\n"
  end

  it 'supports must_raise as to_raise' do
    expect { raise RuntimeError }.to_raise RuntimeError
    expect(->{ raise RuntimeError }).to_raise RuntimeError
  end

  it 'supports must_respond_to as to_respond_to' do
    expect(Object.new).to_respond_to :must_respond_to
  end

  it 'supports wont_respond_to as to_not_respond_to' do
    expect(Object.new).to_not_respond_to :fart_factory
  end

  it 'supports must_throw as to_throw' do
    expect { throw StandardError }.to_throw StandardError
    expect(->{ throw StandardError }).to_throw StandardError
  end
end


wip.rb
old_instance_methods = Object.instance_methods

require 'minitest/spec'

$infected_assertions = Object.instance_methods - old_instance_methods

module Kernel
  def expect object
    Expect.new object
  end
end

class Expect
  attr_reader :object

  def initialize object
    @object = object
  end

  $infected_assertions.each do |assertion|
    method = assertion.to_s.gsub('must', 'to').gsub 'wont', 'to_not'

    self.class_eval <<-EOM
      def #{method} *args
        object.#{assertion} *args
      end
    EOM
  end
end

require 'minitest/autorun'

describe 'forty two' do
  before { @forty_two = 42 }

  it 'equals 42' do
    expect(@forty_two).to_equal 42
  end

  it 'is less than 43' do
    expect(@forty_two).to_be :<, 43
  end

  it 'is not more than 43' do
    expect(@forty_two).to_not_be :>, 43
  end

  it 'is a fixnum' do
    expect(@forty_two).to_be_instance_of Fixnum
  end

  it 'is a kind of integer' do
    expect(@forty_two).to_be_kind_of Integer
  end

  it 'is not a kind of float' do
    expect(@forty_two).to_not_be_kind_of Float
  end
end
